File: solutions/ruby/roman-numerals/1/roman_numerals.rb
class Integer
  def to_roman
    romans = {
      1 => "I",
      4 => "IV",
      5 => "V",
      9 => "IX",
      10 => "X",
      40 => "XL",
      50 => "L",
      90 => "XC",
      100 => "C",
      400 => "CD",
      500 => "D",
      900 => "CM",
      1000 => "M"
    }
    int = self
    numeral = ""
    factors = romans.keys.sort { |x, y| y <=> x }
    factors.each do |factor|
      next if int < factor
      multiple = int / factor
      remainder = int % factor
      numeral << romans[factor] * multiple
      int = remainder
    end
    numeral
  end
end

module BookKeeping
  VERSION = 2
end

File: solutions/go/roman-numerals/1/roman_numerals.go
package romannumerals

import (
	"errors"
	"strings"
)

var RomanNumeralsMap = map[int] string {
	1000 : "M",
	900 : "CM",
	500 : "D",
	400 : "CD",
	100 : "C",
	90 : "XC",
	50 : "L",
	40 : "XL",
	10 : "X",
	9 : "IX",
	5 : "V",
	4 : "IV",
	1 : "I",
}

func ToRomanNumeral(input int) (string, error) {
	if input < 1 {
		return "", errors.New("Roman Numeral must be a positive integer")
	}

	if input > 3999 {
		return "", errors.New("Romans can't count past MMMCMIX")
	}

	var multiple int
	var remainder int

	var b strings.Builder
	for _, factor := range([]int { 1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1}) {
		if input == 0 {
			break
		}
		if input < factor {
			continue
		}

		multiple = input / factor
		remainder = input % factor

		b.WriteString(strings.Repeat(RomanNumeralsMap[factor], multiple))
		input = remainder
	}

	return b.String(), nil
}
